omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+                "s(seconds), m(minutes), h(hours), d(days), w(weeks), M(months) "
+                "e.g 7m = 7 minutes, 1d = 1day, 7m30s = 7 min 30 secs. "
+                "A prefix '_' specifies a negative timedelta. "
+                "This allows for easily getting the setting "
+                "e.g. 2h ago: '_2h' while setting the `time` argument to 'now' (default)."
+            ),
+        ),
+        state=dict(
+            action='store_true',
+            help=(
+                "Prints the state of the StateTracker. "
+                "Does not extract anything else."
+            ),
+        ),
+        output=dict(
+            type=PathOrStr,
+            default='knobs.madx',
+            help=(
+                "Specify user-defined output path. "
+                "This should probably be `model_dir/knobs.madx`"
+            ),
+        ),
+        knob_definitions=dict(
+            type=PathOrStrOrDataFrame,
+            help=(
+                "User defined path to the knob-definitions, "
+                "or (via python) a dataframe containing the knob definitions "
+                "with the columns 'madx', 'lsa' and 'scaling'."
+            ),
+        ),
+    )
+
+
+@dataclass
+class KnobEntry:
+    madx: str
+    lsa: str
+    scaling: float  # is usually +-1, i.e. takes care of sign-conventions
+    value: float = None
+
+    def get_madx(self):
+        if self.value is None:
+            return f"! {self.madx} : No Value extracted"
+        return f"{self.madx} := {self.value * self.scaling};"
+
+
+KnobsDict = Dict[str, KnobEntry]
+
+
+@entrypoint(
+    get_params(), strict=True,
+    argument_parser_args=dict(
+        epilog=USAGE_EXAMPLES,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        prog="Knob Extraction Tool."
+    )
+)
+def main(opt) -> Optional[KnobsDict]:
+    """ Main knob extracting function. """
+    if not opt.time and not opt.state:
+        raise ValueError("No functionality selected. Set either `time` or `state`.")
+
+    ldb = pytimber.LoggingDB(source="nxcals")
+    time = _parse_time(opt.time, opt.timedelta)
+
+    if opt.state:
+        # only print the state of the StateTracker - the MetaState!
+        # I still don't know what this does,
+        # because I only get back that the variable does not exist. (jdilly)
+        state = ldb.get("LhcStateTracker:State", time)  # do first to have output together
+        LOGGER.info("---- STATE ------------------------------------")
+        LOGGER.info(state)
+        return None
+    
+    knobs_dict = _parse_knobs_defintions(opt.knob_definitions)
+    knobs_extract = _extract(ldb, knobs_dict, opt.knobs, time)
+    _write_knobsfile(opt.output, knobs_extract, time)
+    return knobs_extract
+
+
+def _extract(ldb, knobs_dict: KnobsDict, knob_categories: Sequence[str], time: datetime) -> KnobsDict:
+    """
+    Main function to gather data from  the state-tracker.
+
+    Args:
+        ldb: The pytimber database.
```

Comment:
I know the type hint would raise an error in the definition, but we can add `(pytimber.LoggingDB)` here in the docstring.